pkg/api/operators/apis/experiment/v1alpha2/types.go

```diff
@@ -0,0 +1,220 @@
+	ParameterAssignments []trial.ParameterAssignment `json:"parameterAssignments"`
+
+	// Observation for this trial
+	Observation trial.Observation `json:"observation,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+// ExperimentCondition describes the state of the experiment at a certain point.
+type ExperimentCondition struct {
+	// Type of experiment condition.
+	Type ExperimentConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+
+	// A human readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+// ExperimentConditionType defines the state of an Experiment.
+type ExperimentConditionType string
+
+const (
+	ExperimentCreated    ExperimentConditionType = "Created"
+	ExperimentRunning    ExperimentConditionType = "Running"
+	ExperimentRestarting ExperimentConditionType = "Restarting"
+	ExperimentSucceeded  ExperimentConditionType = "Succeeded"
+	ExperimentFailed     ExperimentConditionType = "Failed"
+)
+
+type ParameterSpec struct {
+	Name          string        `json:"name,omitempty"`
+	ParameterType ParameterType `json:"parameterType,omitempty"`
+	FeasibleSpace FeasibleSpace `json:"feasibleSpace,omitempty"`
+}
+
+type ParameterType string
+
+const (
+	ParameterTypeUnknown     ParameterType = "unknown"
+	ParameterTypeDouble      ParameterType = "double"
+	ParameterTypeInt         ParameterType = "int"
+	ParameterTypeDiscrete    ParameterType = "discrete"
+	ParameterTypeCategorical ParameterType = "categorical"
+)
+
+type FeasibleSpace struct {
+	Max  string   `json:"max,omitempty"`
+	Min  string   `json:"min,omitempty"`
+	List []string `json:"list,omitempty"`
+	Step string   `json:"step,omitempty"`
+}
+
+type ObjectiveSpec struct {
+	Type                ObjectiveType `json:"type,omitempty"`
+	Goal                float64       `json:"goal,omitempty"`
+	ObjectiveMetricName string        `json:"objectiveMetricName,omitempty"`
+	// This can be empty if we only care about the objective metric.
+	// Note: If we adopt a push instead of pull mechanism, this can be omitted completely.
+	AdditionalMetricsNames []string `json:"additionalMetricsNames,omitempty"`
+}
+
+type ObjectiveType string
+
+const (
+	ObjectiveTypeUnknown  ObjectiveType = ""
+	ObjectiveTypeMinimize ObjectiveType = "minimize"
+	ObjectiveTypeMaximize ObjectiveType = "maximize"
+)
+
+type AlgorithmSpec struct {
+	AlgorithmName string		 `json:"algorithmName,omitempty"`
+	// Key-value pairs for hyperparameters and assignment values.
+	ParameterAssignments []trial.ParameterAssignment `json:"parameterAssignments"`
```

Comment:
This is algorithm specific parameters.  eg: https://github.com/kubeflow/katib/blob/master/examples/grid-example.yaml#L41

Will this be confusing with ParameterAssignment of Trial though both are key-value struct type?

Reply:
I agree with @johnugeorge . 
In the gRpc API, we distinguish these concepts.
ParameterAssignment of Trial: https://github.com/kubeflow/katib/blob/master/pkg/api/api.proto#L284
Parameter for Suggestion service: https://github.com/kubeflow/katib/blob/master/pkg/api/api.proto#L338

It is also key-value but we should use another name. How about `AlgorithmParameterAssignment` ?

Reply:
I think it should be something like AlgorithmParameter. "Assignment" is for assigning HP values which should not be confused with internal parameters for suggestion algorithms (sorry if my previous answer was misleading).